### PR TITLE
Add auto-gzip contracts in cli when uploading

### PR DESCRIPTION
**Pull request by anilCSE**
Fixes: https://github.com/cosmwasm/wasmd/issues/18

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmwasm/wasmd/blob/master/CONTRIBUTING.md#pr-targeting))

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/wasm/client/utils/utils.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+// limit max bytes read to prevent gzip bombs
+const maxSize = 400 * 1024
+
+// IsGzip returns checks if the file contents are gzip compressed
+func IsGzip(input []byte) bool {
+	if len(input) < 3 {
+		return false
+	}
+
+	in := io.LimitReader(bytes.NewReader(input), maxSize)
```
Comment (ethanfrey) [inline x/wasm/client/utils/utils.go]: Alex does this to maintain a stream for his reader functionality. You can just do a bytes.Equal(input[:3], gzipIdent)

The function is correct and nice code organization btw. Just could be simpler

Comment (ethanfrey) [reply x/wasm/client/utils/utils.go]: Oh, and compare on maxSize as well I guess

File: x/wasm/client/utils/utils.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+		return false
+	}
+
+	return bytes.Equal(gzipIdent, buf)
+}
+
+// IsWasm checks if the file contents are of wasm binary
+func IsWasm(input []byte) bool {
+	if len(input) < 3 {
+		return false
```
Comment (ethanfrey) [inline x/wasm/client/utils/utils.go]: Same as above. Function looks good, but could be simpler.

File: x/wasm/client/utils/utils.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	if bytes.Equal(wasmIdent, buf) {
+		return true
+	}
+
+	return false
+}
+
+// GzipIt compresses the input ([]byte)
+func GzipIt(input []byte) ([]byte, error) {
+	// Create gzip writer.
```
Comment (ethanfrey) [inline x/wasm/client/utils/utils.go]: Nice

File: x/wasm/client/cli/tx.go [hunk 4]
```diff
@@ -57,6 +59,17 @@ func StoreCodeCmd(cdc *codec.Codec) *cobra.Command {
 				return err
 			}
 
+			// gzip the wasm file
+			if wasmUtils.IsWasm(wasm) {
+				wasm, err = wasmUtils.GzipIt(wasm)
```
Comment (ethanfrey) [inline x/wasm/client/cli/tx.go]: Nice

**Comment by anilCSE** [comment]
@ethanfrey updated all the changes. It is r4r

File: x/wasm/client/utils/utils.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	wasmIdent = []byte("\x00\x61\x73\x6D")
+)
+
+// IsGzip returns checks if the file contents are gzip compressed
+func IsGzip(input []byte) bool {
+	return bytes.Equal(input[:3], gzipIdent)
+}
+
+// IsWasm checks if the file contents are of wasm binary
+func IsWasm(input []byte) bool {
```
Comment (ethanfrey) [inline x/wasm/client/utils/utils.go]: Thanks for the cleanup, so simple

**Pull request merged**